src/server/player_sao.cpp

```diff
@@ -463,36 +463,34 @@ void PlayerSAO::rightClick(ServerActiveObject *clicker)
 	m_env->getScriptIface()->on_rightclickplayer(this, clicker);
 }
 
-void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason, bool send)
+void PlayerSAO::setHP(s32 hp, const PlayerHPChangeReason &reason, bool from_client)
 {
-	if (hp == (s32)m_hp)
-		return; // Nothing to do
+	hp = rangelim(hp, 0, U16_MAX);
 
-	if (m_hp <= 0 && hp < (s32)m_hp)
-		return; // Cannot take more damage
+	if (hp == m_hp)
+		return; // Nothing to do
 
-	{
-		s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - m_hp, reason);
-		if (hp_change == 0)
-			return;
+	s32 hp_change = m_env->getScriptIface()->on_player_hpchange(this, hp - (s32)m_hp, reason);
+	hp_change = rangelim(hp_change, -U16_MAX, U16_MAX);
```

Comment:
if it's rangelimited anyway doing this here is pointless

Reply:
Do you mean `hp = rangelim(hp, 0, U16_MAX);` ? Well, on_player_hpchange can return 2147483647 and it will overflow.

Reply:
The `hp_change = rangelim(hp_change, -U16_MAX, U16_MAX);` is pointless.
Sure it can protect against integer overflow but I don't think that's really important here.

Reply:
It won't hurt to protect against stupidness

Reply:
We have the same check there https://github.com/minetest/minetest/blob/1bc753f655db4c7de030f2700361011e1c0278a5/src/script/lua_api/l_item.cpp#L105-L106 for example

Reply:
That's different. Anyway it doesn't matter.